File: backends/java-spring/src/test/java/com/authorization/integration/GateIntegrationTest.java
/**
 * This is a program developed by BobTabo.
 *
 * Copyright (c) 2026 BobTabo. All Rights Reserved.
 */
package com.authorization.integration;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;

import com.authorization.Application;
import java.util.Map;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.server.LocalServerPort;
import org.springframework.test.web.servlet.client.EntityExchangeResult;
import org.springframework.test.web.servlet.client.RestTestClient;

/**
 * Gate（JWT発行・検証）APIの実DB/Redis接続統合テストです。
 * kotlin-ktorの{@code integration/GateIntegrationTest.kt}に相当します。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
@SpringBootTest(classes = Application.class, webEnvironment = RANDOM_PORT)
class GateIntegrationTest {

    @LocalServerPort
    private int port;

    private RestTestClient client;

    @BeforeEach
    void setUp() {
        TestHelper.truncateTables();
        client = RestTestClient.bindToServer().baseUrl("http://localhost:" + port).build();
    }

    @Test
    void issueReturnsJwtTokenForValidClient() {
        var c = TestHelper.createClient();

        EntityExchangeResult<Map> result = client.get()
                .uri("/api/gate/issue?member=user-001")
                .header("Authorization", "Bearer " + c.accessToken())
                .exchange()
                .expectBody(Map.class)
                .returnResult();

        assertThat(result.getStatus().value()).isEqualTo(200);
        assertThat(result.getResponseBody()).containsKey("token");
        assertThat((String) result.getResponseBody().get("token")).isNotBlank();
    }

    @Test
    void issueReturns400WhenMemberParamEmpty() {
        // member は @RequestParam String（required 省略＝必須）のため、パラメーター自体を
        // 省略すると Spring 側のバインディングでコントローラー到達前に400になる。
        // コントローラー自身の member_required 検証を通すため、空文字で明示的に渡す。
        EntityExchangeResult<Map> result = client.get()
                .uri("/api/gate/issue?member=")
                .exchange()
                .expectBody(Map.class)
                .returnResult();

        assertThat(result.getStatus().value()).isEqualTo(400);
        assertThat(result.getResponseBody()).containsEntry("message", "member_required");
    }

    @Test
    void verifyReturnsClaimsForValidToken() {
        var c = TestHelper.createClient();
        EntityExchangeResult<Map> issueResult = client.get()
                .uri("/api/gate/issue?member=user-001")
                .header("Authorization", "Bearer " + c.accessToken())
                .exchange()
                .expectBody(Map.class)
                .returnResult();
        String token = (String) issueResult.getResponseBody().get("token");

        EntityExchangeResult<Map> result = client.get()
                .uri("/api/gate/client/" + c.identifier() + "/verify?token=" + token)
                .exchange()
                .expectBody(Map.class)
                .returnResult();

        assertThat(result.getStatus().value()).isEqualTo(200);
        assertThat(result.getResponseBody()).containsEntry("sub", "user-001");
    }

    @Test
    void verifyReturns400WhenTokenEmpty() {
        // token も @RequestParam String（必須）のため、省略ではなく空文字で明示的に渡し、
        // コントローラー自身の token_required 検証を通す（issueReturns400WhenMemberParamEmpty同様の理由）。
        var c = TestHelper.createClient();

        EntityExchangeResult<Map> result = client.get()
                .uri("/api/gate/client/" + c.identifier() + "/verify?token=")
                .exchange()
                .expectBody(Map.class)
                .returnResult();

        assertThat(result.getStatus().value()).isEqualTo(400);
        assertThat(result.getResponseBody()).containsEntry("message", "token_required");
    }
}

File: backends/kotlin-ktor/src/test/kotlin/com/authorization/integration/GateIntegrationTest.kt
package com.authorization.integration

import com.authorization.module
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.server.testing.*
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlinx.serialization.json.*

class GateIntegrationTest {

    @BeforeTest
    fun setUp() {
        TestHelper.db
        TestHelper.truncateTables()
    }

    @Test
    fun `GET api gate issue returns JWT token`() = testApplication {
        application { module(TestHelper.cfg) }
        val c = TestHelper.createClient()
        val response = client.get("/api/gate/issue?member=user-001") {
            header(HttpHeaders.Authorization, "Bearer ${c.accessToken}")
        }
        assertEquals(HttpStatusCode.OK, response.status)
        val body = Json.parseToJsonElement(response.bodyAsText()).jsonObject
        assertNotNull(body["token"]!!.jsonPrimitive.content)
    }

    @Test
    fun `GET api gate issue returns 400 when member param missing`() = testApplication {
        application { module(TestHelper.cfg) }
        val response = client.get("/api/gate/issue")
        assertEquals(HttpStatusCode.BadRequest, response.status)
        val body = Json.parseToJsonElement(response.bodyAsText()).jsonObject
        assertEquals("member_required", body["error"]!!.jsonPrimitive.content)
    }

    @Test
    fun `GET api gate verify returns claims for valid token`() = testApplication {
        application { module(TestHelper.cfg) }
        val c = TestHelper.createClient()
        val issueResponse = client.get("/api/gate/issue?member=user-001") {
            header(HttpHeaders.Authorization, "Bearer ${c.accessToken}")
        }
        val token = Json.parseToJsonElement(issueResponse.bodyAsText()).jsonObject["token"]!!.jsonPrimitive.content

        val verifyResponse = client.get("/api/gate/client/${c.identifier}/verify?token=$token")
        assertEquals(HttpStatusCode.OK, verifyResponse.status)
        val body = Json.parseToJsonElement(verifyResponse.bodyAsText()).jsonObject
        assertEquals("user-001", body["sub"]!!.jsonPrimitive.content)
    }

    @Test
    fun `GET api gate verify returns 400 when token missing`() = testApplication {
        application { module(TestHelper.cfg) }
        val c = TestHelper.createClient()
        val response = client.get("/api/gate/client/${c.identifier}/verify")
        assertEquals(HttpStatusCode.BadRequest, response.status)
        val body = Json.parseToJsonElement(response.bodyAsText()).jsonObject
        assertEquals("token_required", body["error"]!!.jsonPrimitive.content)
    }
}
